examples/pytorch/correct_and_smooth/main.py

```diff
@@ -0,0 +1,183 @@
+                    activation=F.relu,
+                    dropout=args.dropout,
+                    attn_drop=args.attn_drop)
+    else:
+        raise NotImplementedError(f'Model {args.model} is not supported.')
+
+    model = model.to(device)
+    print(f'Model parameters: {sum(p.numel() for p in model.parameters())}')
+
+    if args.pretrain:
+        print('---------- Before ----------')
+        model.load_state_dict(torch.load(f'base/{args.dataset}-{args.model}.pt'))
+        model.eval()
+
+        if args.model == 'gat':
+            y_soft = model(g, feats).exp()
+        else:
+            y_soft = model(feats).exp()
+
+        y_pred = y_soft.argmax(dim=-1, keepdim=True)
+        valid_acc = evaluate(y_pred, labels, valid_idx, evaluator)
+        test_acc = evaluate(y_pred, labels, test_idx, evaluator)
+        print(f'Valid acc: {valid_acc:.4f} | Test acc: {test_acc:.4f}')
+
+        print('---------- Correct & Smoothing ----------')
+        cs = CorrectAndSmooth(num_correction_layers=args.num_correction_layers,
+                              correction_alpha=args.correction_alpha,
+                              num_smoothing_layers=args.num_smoothing_layers,
+                              smoothing_alpha=args.smoothing_alpha,
+                              scale=args.scale)
+        y_soft = cs.correct(g, y_soft, labels[train_idx], train_idx)
+        y_soft = cs.smooth(g, y_soft, labels[train_idx], train_idx)
+        y_pred = y_soft.argmax(dim=-1, keepdim=True)
+        valid_acc = evaluate(y_pred, labels, valid_idx, evaluator)
+        test_acc = evaluate(y_pred, labels, test_idx, evaluator)
+        print(f'Valid acc: {valid_acc:.4f} | Test acc: {test_acc:.4f}')
+    else:
+        if args.model == 'gat':
+            opt = optim.RMSprop(model.parameters(), lr=args.lr)
+        else:
+            opt = optim.Adam(model.parameters(), lr=args.lr)
+
+        best_acc = 0
+        best_model = copy.deepcopy(model)
+
+        # training
+        print('---------- Training ----------')
+        for i in range(args.epochs):
+            model.train()
+            opt.zero_grad()
+
+            if args.model == 'gat':
+                logits = model(g, feats)
+            else:
+                logits = model(feats)
+            
+            train_loss = F.nll_loss(logits[train_idx], labels.squeeze(1)[train_idx])
+            train_loss.backward()
+
+            opt.step()
+            
+            model.eval()
+            with torch.no_grad():
+                y_pred = logits.argmax(dim=-1, keepdim=True)
+
+                train_acc = evaluate(y_pred, labels, train_idx, evaluator)
+                valid_acc = evaluate(y_pred, labels, valid_idx, evaluator)
+
+                print(f'Epoch {i} | Train loss: {train_loss.item():.4f} | Train acc: {train_acc:.4f} | Valid acc {valid_acc:.4f}')
+
+                if valid_acc > best_acc:
+                    best_acc = valid_acc
+                    best_model = copy.deepcopy(model)
```

Comment:
Best model is evaluated based on best validation loss for GAT: https://github.com/CUAI/CorrectAndSmooth/blob/b910314a59270984f5e249462ee3faa815fc9a0c/gat/gat.py#L162